### ISSUE TITLE
[MIPS] Got `PC offset is too large 0xffffffff....` when linking using LLD

### ISSUE THREAD

**Issue by coder014** [llvm/llvm-project#88852]
When compiling a Rust program to an ELF executable, I encountered the error
```
.....rcgu.o:(.eh_frame): PC offset is too large: 0xffffffff00000968
```
Upon investigation, I found that `lld`, when generating the `.eh_frame_hdr` for DWARF Exception Frame, computes the `pc` as `0xffffffff80020998` and `va` as `0x80020030`. Consequently, `lld` detects an overflow after calculating the offset.
https://github.com/llvm/llvm-project/blob/c13b7485b87909fcf739f62cfa382b55407433c0/lld/ELF/SyntheticSections.cpp#L530-L545

**The program I wrote is a type of kernel code for MIPS32 using static relocation, and I use a linker script to adjust  executable code addresses above `0x80000000`**. Obviously the `pc` obtained by `lld` is incorrect, which overflows the range of a 32-bit integer. Upon reviewing the `getFdePc` code, I found that data reading is based on the **`pointer encoding`** specified in Common Information Entry (CIE). Inspecting the pointer encoding in my object ELF's `cie`, I found it to be **`DW_EH_PE_sdata4`**, causing `lld` to perform **sign extension** when reading the address in FDE.

After further exploration of LLVM MC, I noticed that for MIPS, when not using PIC/PIE, `FDECFIEncoding` is set to `DW_EH_PE_sdata4`,
https://github.com/llvm/llvm-project/blob/c13b7485b87909fcf739f62cfa382b55407433c0/llvm/lib/MC/MCObjectFileInfo.cpp#L338-L352

corroborated by tests.
https://github.com/llvm/llvm-project/blob/c13b7485b87909fcf739f62cfa382b55407433c0/llvm/test/MC/Mips/eh-frame.s#L57-L65

I also attempted linking my object code with `GNU ld`, which did not produce any error messages. It's notable that `GNU ld` imposes additional conditions on overflow checks in `.eh_frame_hdr` generation, not flagging overflow when generating ELF32.
https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=bfd/elf-eh-frame.c;h=902d7c16334ea725b0caac8033d05422316bb4c5;hb=HEAD#l2471
```c
      for (i = 0; i < hdr_info->u.dwarf.fde_count; i++)
	{
	  bfd_vma val;

	  val = hdr_info->u.dwarf.array[i].initial_loc
	    - sec->output_section->vma;
	  val = ((val & 0xffffffff) ^ 0x80000000) - 0x80000000;
	  if (elf_elfheader (abfd)->e_ident[EI_CLASS] == ELFCLASS64
	      && (hdr_info->u.dwarf.array[i].initial_loc
		  != sec->output_section->vma + val))
	    overflow = true;
```

* Should `lld` add additional checks to avoid such errors? 
* For MIPS32 code using static relocation, why we use `DW_EH_PE_sdata4` to fill `FDECFIEncoding` instead of `DW_EH_PE_udata4`, to store absolute addresses? I don't know if this is specified by MIPS ABI or other specifications. If feasible, could we consider to use `DW_EH_PE_udata4` or `DW_EH_PE_absptr`?

**Comment by brad0** [comment]
@MaskRay

**Comment by wzssyqa** [comment]
@coder014 can you attach the objects, so that we can have a test.

```C++
uint64_t EhFrameSection::getFdePc(uint8_t *buf, size_t fdeOff,
                                  uint8_t enc) const {
  // The starting address to which this FDE applies is
  // stored at FDE + 8 byte. And this offset is within
  // the .eh_frame section.
  size_t off = fdeOff + 8;
  uint64_t addr = readFdeAddr(buf + off, enc & 0xf);
  if ((enc & 0x70) == DW_EH_PE_absptr)
    return addr;
  if ((enc & 0x70) == DW_EH_PE_pcrel)
    return addr + getParent()->addr + off + outSecOff;
  fatal("unknown FDE size relative encoding");
}
```
I guess the problem is here.
We should do signed-add for `addr` if `enc` is `DW_EH_PE_pcrel && (DW_EH_PE_sdata4 ||DW_EH_PE_sdata2)`.

**Comment by wzssyqa** [comment]
Can you have a test of this patch?
```
diff --git a/lld/ELF/SyntheticSections.cpp b/lld/ELF/SyntheticSections.cpp
index 7b9ada40c0f6..77ce8c8ab4a9 100644
--- a/lld/ELF/SyntheticSections.cpp
+++ b/lld/ELF/SyntheticSections.cpp
@@ -614,8 +614,15 @@ uint64_t EhFrameSection::getFdePc(uint8_t *buf, size_t fdeOff,
   uint64_t addr = readFdeAddr(buf + off, enc & 0xf);
   if ((enc & 0x70) == DW_EH_PE_absptr)
     return addr;
-  if ((enc & 0x70) == DW_EH_PE_pcrel)
-    return addr + getParent()->addr + off + outSecOff;
+  if ((enc & 0x70) == DW_EH_PE_pcrel) {
+    uint64_t addr_base = getParent()->addr + off + outSecOff;
+    if (enc & 0xf == DW_EH_PE_sdata2)
+      return (int16_t)addr + addr_base;
+    else if (enc & 0xf == DW_EH_PE_sdata4)
+      return (int32_t)addr + addr_base;
+    else
+      return addr + addr_base;
+  }
   fatal("unknown FDE size relative encoding");
 }
```

**Comment by coder014** [comment]
> @coder014 can you attach the objects, so that we can have a test.
> 
> ```c++
> uint64_t EhFrameSection::getFdePc(uint8_t *buf, size_t fdeOff,
>                                   uint8_t enc) const {
>   // The starting address to which this FDE applies is
>   // stored at FDE + 8 byte. And this offset is within
>   // the .eh_frame section.
>   size_t off = fdeOff + 8;
>   uint64_t addr = readFdeAddr(buf + off, enc & 0xf);
>   if ((enc & 0x70) == DW_EH_PE_absptr)
>     return addr;
>   if ((enc & 0x70) == DW_EH_PE_pcrel)
>     return addr + getParent()->addr + off + outSecOff;
>   fatal("unknown FDE size relative encoding");
> }
> ```
> 
> I guess the problem is here. We should do signed-add for `addr` if `enc` is `DW_EH_PE_pcrel && (DW_EH_PE_sdata4 ||DW_EH_PE_sdata2)`.

Here's my minimal working example:
```assembly
# testmips.asm
.text
.set noreorder
.global func
func:
    .cfi_startproc
    .cfi_endproc
```
```
/* testmips.lds */
OUTPUT_ARCH(mips)
ENTRY(func)
SECTIONS
{
    . = 0x80000000;
    .text : { *(.text) }
    .data : { *(.data) }
    .bss : { *(.bss) }
}
```
then assemble them using `llvm-mc` and `ld.lld`:
```bash
llvm-mc -filetype=obj -o testmips.o -arch=mipsel testmips.asm
ld.lld -T testmips.lds -static -o testmips testmips.o
```
`ld.lld: error: testmips.o:(.eh_frame): PC offset is too large: 0xffffffff00000040` is expected.

**Comment by coder014** [comment]
> ```diff
> DW_EH_PE_pcrel
> ```

Thanks for the patch. I tested it but the error still occurs.

Since my program uses **static** linking, there's no `DW_EH_PE_pcrel` flag involved. I don't think the patch will fix the issue.

**Comment by wzssyqa** [comment]
It seems due to assembler, which generates DW_EH_PE_absptr instead of DW_EH_PE_pcrel.
Let's have try to fix it.

**Comment by wzssyqa** [comment]
```diff
diff --git a/lld/ELF/SyntheticSections.cpp b/lld/ELF/SyntheticSections.cpp
index 7b9ada40c0f6..25e49feaaca7 100644
--- a/lld/ELF/SyntheticSections.cpp
+++ b/lld/ELF/SyntheticSections.cpp
@@ -612,10 +612,21 @@ uint64_t EhFrameSection::getFdePc(uint8_t *buf, size_t fdeOff,
   // the .eh_frame section.
   size_t off = fdeOff + 8;
   uint64_t addr = readFdeAddr(buf + off, enc & 0xf);
-  if ((enc & 0x70) == DW_EH_PE_absptr)
-    return addr;
-  if ((enc & 0x70) == DW_EH_PE_pcrel)
-    return addr + getParent()->addr + off + outSecOff;
+  if ((enc & 0x70) == DW_EH_PE_absptr) {
+    if ((enc & 0xf) == DW_EH_PE_sdata2)
+      return (uint64_t)(uint16_t)addr;
+    else if ((enc & 0xf) == DW_EH_PE_sdata4)
+      return (uint64_t)(uint32_t)addr;
+  }
+  if ((enc & 0x70) == DW_EH_PE_pcrel) {
+    uint64_t addr_base = getParent()->addr + off + outSecOff;
+    if ((enc & 0xf) == DW_EH_PE_sdata2)
+      return (int16_t)addr + addr_base;
+    else if ((enc & 0xf) == DW_EH_PE_sdata4)
+      return (int32_t)addr + addr_base;
+    else
+      return addr + addr_base;
+  }
   fatal("unknown FDE size relative encoding");
 }
```

This patch seems working well with your test case.
While I am worrying about that it will break more.
Let's do more test.

**Comment by MaskRay** [comment]
The code snippet above fixes ELFCLASS32 but is probably not desired for ELFCLASS64
```c
+    if ((enc & 0xf) == DW_EH_PE_sdata2)
+      return (uint64_t)(uint16_t)addr;
+    else if ((enc & 0xf) == DW_EH_PE_sdata4)
+      return (uint64_t)(uint32_t)addr;
```

I've created a simpler fix #92438

**Comment by brad0** [comment]
@coder014 Can you please test MaskRay's fix?

**Comment by coder014** [comment]
> @coder014 Can you please test MaskRay's fix?

I tested fix #92438, and it successfully resolves the error. Thanks!

**Comment by brad0** [comment]
> I tested fix #92438, and it successfully resolves the error. Thanks!

Excellent! Thank you.
